File: src/main/java/no/nav/personopplysninger/features/endreopplysninger/domain/telefon/Telefonnummer.java
package no.nav.personopplysninger.features.endreopplysninger.domain.telefon;

import com.fasterxml.jackson.annotation.*;
import com.fasterxml.jackson.core.JsonParseException;
import com.fasterxml.jackson.databind.JsonMappingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;

@JsonIgnoreProperties(ignoreUnknown = true)
public class Telefonnummer {
    @JsonProperty("@type")
    private String subtype = "TELEFONNUMMER";

    private String kilde = "BRUKER SELV";
    private String landskode = null;
    private String nummer;
    private String type;

    public Telefonnummer() {}

    // Denne var nødvendig fordi verdien på feltet 'innmeldtEndring' (som kan representere et Telefonnummer) er
    // en json representert som en string og wrappet med dobbeltfnutter.
    @JsonCreator
    public static Telefonnummer create(String json) throws JsonParseException, JsonMappingException, IOException {
        return (new ObjectMapper()).readValue(json, Telefonnummer.class);
    }

    public Telefonnummer(String kilde, String landskode, String nummer, String type) {
        this.kilde = kilde;
        this.landskode = landskode;
        this.nummer = nummer;
        this.type = type;
    }

    public Telefonnummer(String landskode, String nummer, String type) {
        this.landskode = landskode;
        this.nummer = nummer;
        this.type = type;
    }

    public String getKilde() {
        return kilde;
    }

    public String getLandskode() {
        return landskode;
    }

    public String getNummer() {
        return nummer;
    }

    public String getType() {
        return type;
    }
}

File: src/main/kotlin/no/nav/personopplysninger/features/endreopplysninger/domain/telefon/Telefonnummer.kt
package no.nav.personopplysninger.features.endreopplysninger.domain.telefon

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.core.JsonParseException
import com.fasterxml.jackson.databind.JsonMappingException
import com.fasterxml.jackson.databind.ObjectMapper

import java.io.IOException

@JsonIgnoreProperties(ignoreUnknown = true)
class Telefonnummer {
    @JsonProperty("@type")
    private val subtype = "TELEFONNUMMER"

    var kilde = "BRUKER SELV"
    var landskode: String? = null
        private set
    var nummer: String? = null
        private set
    var type: String? = null
        private set

    constructor() {}

    constructor(kilde: String, landskode: String, nummer: String, type: String) {
        this.kilde = kilde
        this.landskode = landskode
        this.nummer = nummer
        this.type = type
    }

    constructor(landskode: String, nummer: String, type: String) {
        this.landskode = landskode
        this.nummer = nummer
        this.type = type
    }

    companion object {

        // Denne var nødvendig fordi verdien på feltet 'innmeldtEndring' (som kan representere et Telefonnummer) er
        // en json representert som en string og wrappet med dobbeltfnutter.
        @JsonCreator
        @JvmStatic
        @Throws(JsonParseException::class, JsonMappingException::class, IOException::class)
        fun create(json: String): Telefonnummer {
            return ObjectMapper().readValue(json, Telefonnummer::class.java)
        }
    }
}
